openam-shared/src/main/java/com/sun/identity/shared/encode/CookieUtils.java

```diff
@@ -474,10 +476,17 @@ public static void addCookieToResponse(HttpServletResponse response, Cookie cook
         Set<String> domains = new HashSet<>();
 
         for (String domain : cookieDomains) {
-            if (domain == null || host.contains(domain)) {
+            if (domain != null && StringUtils.endsWithIgnoreCase(host, domain)) {
                 domains.add(domain);
             }
         }
+        //IE fix: allow add .app.domain.com after .domain.com without SystemPropertiesManager.getAsBoolean(Constants.SET_COOKIE_TO_ALL_DOMAINS, true) 
```

Comment:
If you want to retreive cookies in particular order, you should use LinkedHashSet instand of HashSet